### PR TITLE
Add a <partial /> tag helper

### LINKED ISSUE #5916: Add a <partial /> Tag Helper
Right now, there are four methods on `IHtmlHelper` for rendering partials, and it isn't clear to users if one is *better* than the other. `RenderPartialAsync` is certainly much harder to call and looks much uglier at the call site.

``` HTML
@Html.Partial("_AnalyticsBody")
@await  Html.PartialAsync("_AnalyticsBody")
@{ Html.RenderPartial("_AnalyticsBody"); }
@{ await Html.RenderPartialAsync("_AnalyticsBody"); }
```

All of these methods also have an overload that accepts a model object, e.g. `@await  Html.PartialAsync("_AnalyticsBody", SomeModel)`

It would be much nicer to have a Tag Helper for partial rendering that just does "the right thing":
``` HTML
<partial name="_AnalyticsBody" model="SomeModel" />
```

Strawman implementation:
``` c#
namespace Microsoft.AspNetCore.Mvc.Razor.TagHelpers
{
    [HtmlTargetElement("partial", Attributes = "name")]
    public class PartialTagHelper : TagHelper
    {
        private readonly IHtmlHelper _htmlHelper;

        public PartialTagHelper(IHtmlHelper htmlHelper)
        {
            _htmlHelper = htmlHelper;
        }

        [ViewContext]
        public ViewContext ViewContext { get; set; }

        public string Name { get; set; }

        public object Model { get; set; }

        public override async Task ProcessAsync(TagHelperContext context, TagHelperOutput output)
        {
            ((IViewContextAware)_htmlHelper).Contextualize(ViewContext);

            // Determine best way to do this, e.g. calling RenderPartialCoreAsync() instead passing a TextWriter that does the right thing if that's actually better, handle null Model, etc.
            var content = await _htmlHelper.PartialAsync(Name, Model);
            
            output.TagName = null;
            output.Content.SetHtmlContent(content);
        }
    }
}
```

@NTaylorMullen @rynowak @pranavkm @Eilon

### PR DESCRIPTION
**Pull request by pranavkm**
Fixes #5916

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs [hunk 1]
```diff
@@ -1095,6 +1071,65 @@ public virtual IHtmlContent GenerateAntiforgery(ViewContext viewContext)
...
+            var view = viewEngineResult.View;
+            using (view as IDisposable)
+            {
+                // Determine which ViewData we should use to construct a new ViewData
+                var baseViewData = viewData ?? viewContext.ViewData;
+
+                var newViewData = new ViewDataDictionary<object>(baseViewData, model);
+                var partialViewContext = new ViewContext(viewContext, view, newViewData, writer);
+
+                await viewEngineResult.View.RenderAsync(partialViewContext);
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs]: await view.RenderAsync(...)

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/IHtmlGenerator.cs [hunk 2]
```diff
@@ -481,5 +483,26 @@ public interface IHtmlGenerator
             ModelExplorer modelExplorer,
             string expression,
             bool allowMultiple);
+
+        /// <summary>
+        /// Renders HTML markup for the specified partial view.
+        /// </summary>
+        /// <param name="partialViewName">
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/IHtmlGenerator.cs]: Wrong order. Should be below ViewContext.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/IHtmlGenerator.cs [hunk 3]
```diff
@@ -481,5 +483,26 @@ public interface IHtmlGenerator
...
+        /// The name of the partial view used to create the HTML markup. Must not be <c>null</c>.
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/IHtmlGenerator.cs]: Should probably remove `Must not be null` here or add it to both `ViewContext` and `writer` because we null check those.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewFeatures/DefaultHtmlGeneratorTest.cs [hunk 4]
```diff
@@ -676,8 +677,58 @@ public void GenerateAntiforgery_AlwaysGeneratesAntiforgeryToken_IfCannotRenderAt
...
+                "location4";
+
+            var viewEngine = new Mock<ICompositeViewEngine>(MockBehavior.Strict);
+            viewEngine
+                .Setup(v => v.GetView(/*executingFilePath*/ null, It.IsAny<string>(), /*isMainPage*/ false))
+                .Returns(ViewEngineResult.NotFound("test-view", new[] { "location1", "location2" }))
+                .Verifiable();
+            viewEngine
+                .Setup(v => v.FindView(It.IsAny<ActionContext>(), It.IsAny<string>(), /*isMainPage*/ false))
+                .Returns(ViewEngineResult.NotFound("test-view", new[] { "location3", "location4" }))
```
Comment (ajaybhargavb) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewFeatures/DefaultHtmlGeneratorTest.cs]: I assume we have a test somewhere else where `GetView` fails but `FindView` succeeds? If not, it might be worth adding one.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/breakingchanges.netcore.json [hunk 5]
```diff
@@ -0,0 +1,12 @@
+﻿[
+  {
+    "TypeId": "public class Microsoft.AspNetCore.Mvc.ViewFeatures.DefaultHtmlGenerator : Microsoft.AspNetCore.Mvc.ViewFeatures.IHtmlGenerator",
+    "MemberId": "public .ctor(Microsoft.AspNetCore.Antiforgery.IAntiforgery antiforgery, Microsoft.Extensions.Options.IOptions<Microsoft.AspNetCore.Mvc.MvcViewOptions> optionsAccessor, Microsoft.AspNetCore.Mvc.ModelBinding.IModelMetadataProvider metadataProvider, Microsoft.AspNetCore.Mvc.Routing.IUrlHelperFactory urlHelperFactory, System.Text.Encodings.Web.HtmlEncoder htmlEncoder, Microsoft.AspNetCore.Mvc.ViewFeatures.ValidationHtmlAttributeProvider validationAttributeProvider)",
+    "Kind": "Removal"
```
[the remark this answers is not in the document]
Comment (Eilon) [reply src/Microsoft.AspNetCore.Mvc.ViewFeatures/breakingchanges.netcore.json]: Yeah no breaking changes please.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+            using (var writer = new ViewBufferTextWriter(viewBuffer, Encoding.UTF8))
+            {
+                await _htmlGenerator.RenderPartialViewAsync(
+                    ViewContext,
+                    Name,
+                    Model,
+                    ViewData,
+                    writer);
+
+                output.SuppressOutput();
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs]: This will nuke any other content added by other `TagHelper`s. If someone wants to add anything to the partial `TagHelper` i'd say let them. Instead of suppressing output just set the `TagName` to null to work nicely with anyone else.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        /// </summary>
+        public object Model { get; set; }
+
+        /// <summary>
+        /// A <see cref="ViewDataDictionary"/> to pass into the partial view.
+        /// </summary>
+        public ViewDataDictionary ViewData { get; set; }
+
+        [ViewContext]
+        public ViewContext ViewContext { get; set; }
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs]: `[HtmlAttributeNotBound]`

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+                    Resources.FormatViewEngine_PartialViewNotFound(Name, locations));
+            }
+
+            var view = viewEngineResult.View;
+            using (view as IDisposable)
+            {
+                // Determine which ViewData we should use to construct a new ViewData
+                var baseViewData = ViewData ?? ViewContext.ViewData;
+
+                var newViewData = new ViewDataDictionary<object>(baseViewData, Model);
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs]: Overloads of the `PartialAsync(...)` and `RenderPartialAsync(...)` helpers default the model they use to `htmlHelper.ViewData.Model`. This instead defaults to `null`, which is unexpected. Should do something similar to the HTML helpers here.

Always use `ViewContext.ViewData.Model` since setting the `ViewData` property should not change the `Model` property's default.

File: test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/PartialTagHelperTest.cs [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+                .Returns(Task.CompletedTask);
+
+            var viewEngine = new Mock<ICompositeViewEngine>();
+            viewEngine.Setup(v => v.GetView(It.IsAny<string>(), partialName, false))
+                .Returns(ViewEngineResult.Found(partialName, view.Object));
+
+            var tagHelper = new PartialTagHelper(viewEngine.Object, bufferScope)
+            {
+                Name = partialName,
+                Model = model,
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/PartialTagHelperTest.cs]: Add tests where `Model` (now `For`) isn't set or is explicitly set to `null` but a view is found.

File: test/WebSites/HtmlGenerationWebSite/Views/HtmlGeneration_Home/ProductListUsingTagHelpers.cshtml [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+</head>
+<body>
+    <form asp-action="Index" asp-controller="HtmlGeneration_Product" asp-antiforgery="true">
+    @{
+        var index = 0;
+        var fieldPrefix = ViewData.TemplateInfo.HtmlFieldPrefix;
+        foreach (var product in Model)
+        {
+            @* Update HtmlFieldPrefix so generated for, id and name attribute values are correct. *@
+            ViewData.TemplateInfo.HtmlFieldPrefix = fieldPrefix + string.Format("[{0}]", index++);
```
Comment (dougbu) [inline test/WebSites/HtmlGenerationWebSite/Views/HtmlGeneration_Home/ProductListUsingTagHelpers.cshtml]: This wouldn't be necessary when using `asp-for="[index++]"`.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+            {
+                await RenderPartialViewAsync(writer);
+
+                // Reset the TagName. We don't want `partial` to render.
+                output.TagName = null;
+                output.Content.SetHtmlContent(viewBuffer);
+            }
+        }
+
+        private async Task RenderPartialViewAsync(TextWriter writer)
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs]: Just a suggestion: Could make this tag helper an `HtmlHelper` subclass which implements `ITagHelper`. Would have to make `ViewContext` a `new` property that calls `Contextualize(...)` in its setter but that should be the primary wrinkle.

Comment (dougbu) [reply src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs]: Forgot the punchline: Then could use `RenderPartialCoreAsync(...)`.

Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs]: The sample @DamianEdwards has essentially DIs a `IHtmlHelper`, `Contextualizes` it and then renders the view using it. That kinda felt odd since we don't have any other use cases of doing this in any other tag helper. Given we already have a similarish copy of the code in PartialViewExecutor, I didn't think it was terrible to have yet another copy of it here.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+
+        /// <summary>
+        /// The name of the partial view used to create the HTML markup.
+        /// </summary>
+        public string Name { get; set; }
+
+        /// <summary>
+        /// A model to pass into the partial view.
+        /// </summary>
+        public object Model { get; set; }
```
[the remark this answers is not in the document]
Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs]: I checked with @DamianEdwards and he doesn't think this should be a ModelExpression. He was fine leaving this property as is.

**Comment by pranavkm** [comment]
🆙 📅

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+        public PartialTagHelper(
+            ICompositeViewEngine viewEngine,
+            IViewBufferScope viewBufferScope)
+        {
+            _viewEngine = viewEngine ?? throw new ArgumentNullException(nameof(viewEngine));
+            _viewBufferScope = viewBufferScope ?? throw new ArgumentNullException(nameof(viewBufferScope));
+        }
+
+        /// <summary>
+        /// The name of the partial view used to create the HTML markup.
```
Comment (Eilon) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs]: Does this do just view "names" (e.g. `_LoginPartial`) or does it also do paths? Do we need to be clear about this in the docs?

Comment (Eilon) [reply src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs]: I didn't see any tests using a path, but would it work? And if paths aren't supported:
1. Should they be?
2. The docs need to be clearer about what a "partial view name" is (perhaps by giving an example)

Comment (dougbu) [reply src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs]: Looks like these properties are handled identically to e.g. `viewName` parameters in `Controller.View(...)` methods. Those are generally documented as `The name of the view that is rendered to the response.`. Same for the corresponding result properties e.g. `ViewResult.ViewName`.

If we need more documentation about what names, paths, et cetera mean, suggest we handle that in a separate issue. For now, let's be consistent.

Comment (pranavkm) [reply src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs]: It does path - https://github.com/aspnet/Mvc/pull/7089/files#diff-883159b6acb104f0fb64afa28c27a9ccR2 . The name I think comes from the IHtmlHelper docs. I'd be happy to update it if you have suggestions

Comment (dougbu) [reply src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs]: Oh, and "partial view name" doesn't appear anywhere. Isn't "name of the partial view" clear enough (assuming we stick with just "name" for now)?

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+                // Determine which ViewData we should use to construct a new ViewData
+                var baseViewData = ViewData ?? ViewContext.ViewData;
+                var model = For?.Model ?? ViewContext.ViewData.Model;
+                var newViewData = new ViewDataDictionary<object>(baseViewData, model);
+                var partialViewContext = new ViewContext(ViewContext, view, newViewData, writer);
+
+                if (For?.Name != null)
+                {
+                    newViewData.TemplateInfo.HtmlFieldPrefix = newViewData.TemplateInfo.GetFullHtmlFieldName(For.Name);
+                }
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs]: Could move everything except `var partialViewContext = ...;` outside the `using`. Up to you.

File: src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs (hunk 13 above)
Comment (Eilon) [reply src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs]: > For now, let's be consistent.

I'd rather have correct information than have consistent-but-useless information

> Oh, and "partial view name" doesn't appear anywhere.

These mean the same thing so I don't think this is critical.

...

So, let's come up with some good wording here and then apply elsewhere.

Comment (dougbu) [reply src/Microsoft.AspNetCore.Mvc.TagHelpers/PartialTagHelper.cs]: How 'bout "The name or path of the [partial] view that is rendered to the response." We do lots of shenanigans under the covers handling absolute and relative paths as well as distinguishing paths from names. But, describing exactly how we choose to interpret "name" versus "path" doesn't help much -- especially because it's specific to the `RazorViewEngine`.

I'm not positive what extra words we need to inform readers what distinguishes a "partial" view. It's a view found with `isMainPage: false` but the interfaces say only that `isMainPage` "Determines if the page being found is the main page for an action.". The details are again specific to the `IViewEngine` implementation. (For `RazorViewEngine`, this parameter mainly controls whether or not `_ViewStart.cshtml` files are used.)
